graphql/src/execution/query.rs

```diff
@@ -23,6 +25,41 @@ use crate::{
     schema::api::ErrorPolicy,
 };
 
+lazy_static! {
+    static ref GRAPHQL_VALIDATION_PLAN: ValidationPlan = ValidationPlan::from(
+        if std::env::var("DISABLE_GRAPHQL_VALIDATIONS")
+            .unwrap_or_else(|_| "false".into())
+            .parse::<bool>()
+            .unwrap_or_else(|_| false)
+        {
+            vec![]
+        } else {
+            vec![
+                Box::new(UniqueOperationNames {}),
+                Box::new(LoneAnonymousOperation {}),
+                Box::new(SingleFieldSubscriptions {}),
+                Box::new(KnownTypeNames {}),
+                Box::new(FragmentsOnCompositeTypes {}),
+                Box::new(VariablesAreInputTypes {}),
+                Box::new(LeafFieldSelections {}),
+                Box::new(FieldsOnCorrectType {}),
+                Box::new(UniqueFragmentNames {}),
+                Box::new(KnownFragmentNames {}),
+                Box::new(NoUnusedFragments {}),
+                Box::new(OverlappingFieldsCanBeMerged {}),
+                Box::new(NoFragmentsCycle {}),
```

Comment:
This check needs to be the first one that runs, otherwise other checks that follow fragments can recurse indefinitely.

Actually, looking at the code in `graphql-tools-rs`, `validate` always runs all the validations in the plan, but if fragments contain cycles, we must not ever do anything that recurses on fragments. One simple fix for now would be to call `validate` _after_ the call to `check_complexity` - that checks for cycles and returns an error.

Reply:
With that change, the `NoFragmentsCycle` check is actually redundant and should be removed from the plan.

Reply:
Fixed, and added integration tests to `graphql-tools` to make sure none of the rules are having infinite loops in any rule. 

The GraphQL validation rules are running all validations and do not short-circuit, since it needs to check all rules before returning a result.

The `QueryVisitor` only visits the tree and doesn't resolve `FragmentSpread` so this is something that each rule can do based on its needs.

Reply:
Also, see `core/tests/interfaces.rs` changes - it's now capturing the cyclic fragments using the validation phase - so if the validation will have an infinite loop - those tests will fail :)